rllib/algorithms/dt/dt_torch_policy.py

```diff
@@ -0,0 +1,553 @@
+import gym
+import numpy as np
+
+from typing import (
+    Dict,
+    List,
+    Tuple,
+    Type,
+    Union,
+    Optional,
+    Any,
+    TYPE_CHECKING,
+)
+
+import tree
+from gym.spaces import Discrete, Box
+
+from ray.rllib.algorithms.dt.dt_torch_model import DTTorchModel
+from ray.rllib.evaluation.postprocessing import discount_cumsum
+from ray.rllib.models.catalog import ModelCatalog
+from ray.rllib.models.modelv2 import ModelV2
+from ray.rllib.models.torch.mingpt import configure_gpt_optimizer
+from ray.rllib.models.torch.torch_action_dist import (
+    TorchDistributionWrapper,
+    TorchCategorical,
+    TorchDeterministic,
+)
+from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
+from ray.rllib.policy.torch_mixins import LearningRateSchedule
+from ray.rllib.policy.torch_policy_v2 import TorchPolicyV2
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.annotations import override, PublicAPI
+from ray.rllib.utils.numpy import convert_to_numpy
+from ray.rllib.utils.torch_utils import apply_grad_clipping
+from ray.rllib.utils.typing import (
+    TrainerConfigDict,
+    TensorType,
+    TensorStructType,
+)
+
+if TYPE_CHECKING:
+    from ray.rllib.evaluation import Episode  # noqa
+
+torch, nn = try_import_torch()
+F = nn.functional
+
+
+class DTTorchPolicy(LearningRateSchedule, TorchPolicyV2):
+    def __init__(
+        self,
+        observation_space: gym.spaces.Space,
+        action_space: gym.spaces.Space,
+        config: TrainerConfigDict,
+    ):
+        LearningRateSchedule.__init__(
+            self,
+            config["lr"],
+            config["lr_schedule"],
+        )
+
+        TorchPolicyV2.__init__(
+            self,
+            observation_space,
+            action_space,
+            config,
+            max_seq_len=config["model"]["max_seq_len"],
+        )
+
+    @override(TorchPolicyV2)
+    def make_model_and_action_dist(
+        self,
+    ) -> Tuple[ModelV2, Type[TorchDistributionWrapper]]:
+        # Model
+        model_config = self.config["model"]
+        # TODO: make these better with better AlgorithmConfig options.
+        model_config.update(
+            dict(
```

Comment:
**TODO**: you can pass kwargs directly to dict.update()

Reply:
resolved